src/data_structure/graph.rs

```diff
@@ -0,0 +1,221 @@
+use std::collections::HashMap;
+use std::collections::HashSet;
+use std::marker::PhantomData;
+
+/// 'Graph<N, E, Ty>' is a data structure for graphs with optionally
+/// weighted nodes and edges of arbitrary type. This implementation
+/// is inteded for educational purposes only. 
+/// See [petgraph](https://docs.rs/crate/petgraph/) for a full featured
+/// implementation.
+pub struct Graph<N, E, Ty = Directed> {
+    nodes: HashMap<usize, Node<N>>,
+    edges: HashMap<usize, Edge<E>>,
+    ty: PhantomData<Ty>,
+}
+
+impl<N, E, Ty> Graph<N, E, Ty>
+where
+    Ty: EdgeType,
+{
+    fn new() -> Self {
+        Graph {
+            nodes: HashMap::new(),
+            edges: HashMap::new(),
+            ty: PhantomData,
+        }
+    }
+
+    /// Returns true iff graph is directed.
+    fn is_directed(&self) -> bool {
+        Ty::is_directed()
+    }
+
+    /// Add node to graph and return its assigned index.
+    fn add_node(&mut self, weight: N) -> usize {
+        let node = Node { weight: weight };
+        let mut index = 0usize;
+        while self.nodes.contains_key(&index) {
```

Comment:
This while loops would be inefficient here. Can we do something like `self.nodes.insert(self.nodes.len(), node)`?

Reply:
Yes, thanks for the suggestion. As far as I can tell there is no reason to assign the smallest available index to a new node or edge.

Reply:
Ive thought about it and
`self.nodes.insert(self.nodes.len(), node)`
will not work if we delete and insert a bunch of nodes (i.e. it will result in duplicate usages of a certain index). But it still is the most reasonable first attempt for an index and by decreasing it, if necessary, we will always find an available index, i.e. 
`let mut index = self.nodes.len();
while self.nodes.contains_key(&index) {
            index -= 1;
        }`
will always result in an available index and, most of the time, it will do so in one step.

Reply:
I think we can use something like "Queue" or "Stack" to maintain a list of avaliable indices. Every time a node/edge was deleted, add the delected index to that free list. When we want to add a item, we pop one index from the free list. If that list is empy, then we use `self.xxx.len()` as the index.

I have write some code: <https://github.com/plus7wist/Rust/blob/d5039cee609df8d2823ddab01df40fa22febb2a6/src/data_structures/graph.rs#L280>

Reply:
> The indices, in my mind, are purely for bookkeeping purposes. If a user wants to add labels to their nodes (edges), they should add them to their weights. So, a weighted graph with labels may have weights of type (String, i32), for example.
> 
> If this agrees with you, I'd suggest renaming 'weight' into 'data' or 'payload' to reflect this more general purpose.
> 
> But I'm open to suggestions on this.

Well, if we add our label into `weight`, we can't find a node/edge by our label effectively. In this problem, I agree with @AnshulMalik more.